src/Microsoft.ML.TimeSeries/SrCnnEntireTransformer.cs

```diff
@@ -0,0 +1,903 @@
+                for (int i = 0; i < count; ++i)
+                {
+                    batchResult.Add(valList.GetRange(i * length, length).ToArray());
+                }
+                _anomalyDict = trainingTimestamp.Zip(batchResult, (k, v) => new { k, v }).ToDictionary(x => x.k, x => x.v);
+            }
+
+            void ICanSaveModel.Save(ModelSaveContext ctx) => SaveModel(ctx);
+
+            private void SaveModel(ModelSaveContext ctx)
+            {
+                _host.CheckValue(ctx, nameof(ctx));
+                ctx.CheckAtModel();
+                ctx.SetVersionInfo(GetVersionInfo());
+
+                ctx.Writer.Write((Int32)_parent.OutputLength);
+                ctx.Writer.Write((Int32)_anomalyDict.Count);
+
+                List<long> keyList = new List<long>();
+                List<double> valList = new List<double>();
+                foreach(var pair in _anomalyDict)
+                {
+                    keyList.Add(pair.Key);
+                    valList.AddRange(pair.Value);
+                }
+                ctx.Writer.WriteLongStream(keyList);
+                ctx.Writer.WriteDoubleStream(valList);
+            }
+
+            public void Train(IDataView data) => Train(new RoleMappedData(data, null, _parent.InputColumnName));
+
+            public void Train(RoleMappedData data)
+            {
+                _host.CheckValue(data, nameof(data));
+                _host.CheckParam(data.Schema.Feature.HasValue, nameof(data), "Must have features column.");
+                var featureCol = data.Schema.Feature.Value;
+                if (!(featureCol.Type is SrCnnTsPointDataViewType))
+                    throw _host.ExceptSchemaMismatch(nameof(data), "feature", featureCol.Name, "SrCnnTsPointDataViewType", featureCol.Type.ToString());
+
+                List<Double> trainingData = new List<Double>();
+                List<long> trainingTimestamp = new List<long>();
+                DateTime prevTimestamp = DateTime.MinValue;
+                using (var cursor = data.Data.GetRowCursor(featureCol))
+                {
+                    var getVal = cursor.GetGetter<SrCnnTsPoint>(featureCol);
+                    SrCnnTsPoint val = null;
+                    while (cursor.MoveNext())
+                    {
+                        getVal(ref val);
+                        if (val != null && val.Timestamp > prevTimestamp && !Double.IsNaN(val.Value))
+                        {
+                            trainingTimestamp.Add(val.Timestamp.Ticks);
+                            trainingData.Add(val.Value);
+                            prevTimestamp = val.Timestamp;
+                        }
+                    }
+                }
+
+                _host.CheckIO(trainingTimestamp.Count >= _minTrainingPoint, "Input must have no less than 12 valid points to fit a model.");
+
+                if (trainingTimestamp.Count >= _minTrainingPoint)
+                {
```

Comment:
You are throwing an exception above if `trainingTimeStamp.Count` is less than `_minTrainingPoint`.

Either you don't need this if check here or if you keep the check here, you will need an else block and can move the exception to the else block

Reply:
> You are throwing an exception above if `trainingTimeStamp.Count` is less than `_minTrainingPoint`.
> 
> Either you don't need this if check here or if you keep the check here, you will need an else block and can move the exception to the else block

Updated as you suggested.